pyuvdata/uvbeam.py

```diff
@@ -900,14 +960,134 @@ def get_lambda(real_lut, imag_lut=None):
+                                          'vectors that are not aligned to the '
+                                          'native theta/phi coordinate system '
+                                          'is not yet supported')
+            else:
+                """ The basis vector array comes in defined at the rectangular grid.
+                Redefine it for the interpolation points """
+                interp_basis_vector = np.zeros([self.Naxes_vec,
+                                                self.Ncomponents_vec,
+                                                npoints])
+                interp_basis_vector[0, 0, :] = np.ones(npoints)  # theta hat
+                interp_basis_vector[1, 1, :] = np.ones(npoints)  # phi hat
+        else:
+            interp_basis_vector = None
+
+        for index1 in range(self.Nspws):
+            for index3 in range(input_nfreqs):
+                freq = freq_array[index3]
+                for index0 in range(self.Naxes_vec):
+                    for index2 in range(Npol_feeds):
+                        if np.iscomplexobj(input_data_array):
+                            # interpolate real and imaginary parts separately
+                            real_hmap = hp.get_interp_val(input_data_array[index0, index1, pol_inds[index2], index3, :].real, za_array, az_array)
+                            imag_hmap = hp.get_interp_val(input_data_array[index0, index1, pol_inds[index2], index3, :].imag, za_array, az_array)
+                            hmap = real_hmap + 1j * imag_hmap
+                        else:
+                            # interpolate once
+                            hmap = hp.get_interp_val(input_data_array[index0, index1, pol_inds[index2], index3, :], za_array, az_array)
+                            
+                        interp_data[index0, index1, index2, index3, :] = hmap
+
+        return interp_data, interp_basis_vector
+
+    def interp(self, az_array=None, za_array=None, freq_array=None, freq_interp_tol=1.0,
+               polarizations=None, **kwargs):
         """
         Interpolate beam to given az, za locations (in radians).
 
         Args:
             az_array: az values to interpolate to (same length as za_array)
             za_array: za values to interpolate to (same length as az_array)
             freq_array: frequency values to interpolate to
+            freq_interp_tol: float, frequency distance tolerance [Hz] of nearest neighbors.
+                If *all* elements in freq_array have nearest neighbor distances within
+                the specified tolerance then return the beam at each nearest neighbor,
+                otherwise interpolate the beam.
+            polarizations: list of str, polarizations to interpolate if beam_type is 'power'.
+                Default is all polarizations in self.polarization_array.
+            kwargs: dictionary of keyword arguments to pass to interpolation function. See it for details.
```

Comment:
So the problem with this docstring, is that the interpolation functions are private functions (starting with the underscore character), which means that their docstrings are not part of the public API and are not printed in the ReadTheDocs auto documentation. We made those private because we didn't want users to call those functions directly, we wanted them to call this `interp` method. I think therefore that we need to have the keywords that only apply to specific interpolation methods documented here. I think I'd really prefer to pass those with their proper names and raise an error or warning if they do not apply to the chosen interpolation function. That approach is much more transparent to the user than passing them via `kwargs`. I'm happy to help with this if you like. It looks to me like the only keywords that this applies to currently are `reuse_spline` and `new_object`, am I missing something?

Reply:
Sure thing!